gc.c

```diff
@@ -1884,6 +1884,52 @@ rb_gc_initial_stress_set(VALUE flag)
     initial_stress = flag;
 }
 
+static void * Alloc_GC_impl(void);
+
+#if USE_SHARED_GC
+# include "dln.h"
+# define Alloc_GC rb_gc_functions->init
+
+void
+ruby_external_gc_init()
+{
+    rb_gc_function_map_t *map = malloc(sizeof(rb_gc_function_map_t));
+    rb_gc_functions = map;
+
+    char *gc_so_path = getenv("RUBY_GC_LIBRARY_PATH");
+    if (!gc_so_path) {
+        map->init = Alloc_GC_impl;
+        return;
+    }
+
+    void *h = dln_open(gc_so_path);
+    if (!h) {
+        rb_bug(
+            "ruby_external_gc_init: Shared library %s cannot be opened.\n",
+            gc_so_path
+        );
+    }
+
+    void *gc_init_func = dln_symbol(h, "Init_GC");
+    if (!gc_init_func) {
+        rb_bug("ruby_external_gc_init: Init_GC func not exported by library");
```

Comment:
The failing shared library path is not needed?

```suggestion
        rb_bug("ruby_external_gc_init: Init_GC func not exported by library %s.",
            gc_so_path);
```